providers/implementations/ciphers/cipher_enull_hmac.c

```diff
@@ -0,0 +1,343 @@
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_IVLEN, NULL),
+    OSSL_PARAM_END
+};
+const OSSL_PARAM *enull_hmac_settable_ctx_params(ossl_unused void *cctx,
+                                                 ossl_unused void *provctx)
+{
+    return enull_hmac_known_settable_ctx_params;
+}
+
+static int enull_hmac_einit(void *vctx,
+                            const unsigned char *key, size_t keylen,
+                            const unsigned char *iv, size_t ivlen,
+                            const OSSL_PARAM params[])
+{
+    int ret;
+
+    /* The generic function checks for ossl_prov_is_running() */
+    ret = ossl_cipher_generic_einit(vctx, key, keylen, iv, ivlen, NULL);
+    if (ret && iv != NULL) {
+        PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+        PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+        ret = hw->initiv(ctx, iv, ivlen);
+    }
+    if (ret && !enull_hmac_set_ctx_params(vctx, params))
+        ret = 0;
+    return ret;
+}
+
+static int enull_hmac_dinit(void *vctx,
+                            const unsigned char *key, size_t keylen,
+                            const unsigned char *iv, size_t ivlen,
+                            const OSSL_PARAM params[])
+{
+    int ret;
+
+    /* The generic function checks for ossl_prov_is_running() */
+    ret = ossl_cipher_generic_dinit(vctx, key, keylen, iv, ivlen, NULL);
+    if (ret && iv != NULL) {
+        PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+        PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+        hw->initiv(ctx, iv, ivlen);
+    }
+    if (ret && !enull_hmac_set_ctx_params(vctx, params))
+        ret = 0;
+    return ret;
+}
+
+static int enull_hmac_cipher(void *vctx, unsigned char *out,
+                             size_t *outl, size_t outsize,
+                             const unsigned char *in, size_t inl)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (inl == 0) {
+        *outl = 0;
+        return 1;
+    }
+
+    if (outsize < inl) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    if (!hw->base.cipher(ctx, out, in, inl))
+        return 0;
+    *outl = inl;
+    return 1;
+}
+
+static int enull_hmac_final(void *vctx, unsigned char *out, size_t *outl,
+                            size_t outsize)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    PROV_CIPHER_HW_ENULL_HMAC *hw = (PROV_CIPHER_HW_ENULL_HMAC *)ctx->hw;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (!hw->base.cipher(ctx, out, NULL, 0))
+        return 0;
+
+    *outl = 0;
+    return 1;
+}
+
+#define IMPLEMENT_cipher(lcmd, UCMD, flags, kbits, blkbits, ivbits)            \
+static OSSL_FUNC_cipher_get_params_fn enull_hmac_##lcmd##_get_params;          \
```

Comment:
Why this forward declaration, given the following line actually defining `enull_hmac_##lcmd##_get_params`?

Reply:
It is not forward declaration, `OSSL_FUNC_cipher_get_params_fn ` is macro for typedef function signature.

Reply:
BTW, I could not find where `OSSL_FUNC_cipher_get_params_fn` is defined.

Yet anyway, this is not what I meant.
As I wrote, 
```
static OSSL_FUNC_cipher_get_params_fn enull_hmac_##lcmd##_get_params;
```
is just a forward declaration for the function defined in the subsequent line:
```
static int enull_hmac_##cmd##_get_params(OSSL_PARAM params[])
```
and it is redundant - one can simply leave out
```
static OSSL_FUNC_cipher_get_params_fn enull_hmac_##lcmd##_get_params;`
```

The same BTW holds for many `static OSSL_FUNC_..._fn ...;` decls in `providers/implementations/ciphers/*.c`
Their only effect is to make sure the compiler throws an error if the function definition in the next line by mistake does not have the right type.

Reply:
Yes, I agree it is used just to make sure implemented functions are as per the defined signature. 
It is defined my helper macro `OSSL_CORE_MAKE_FUNC` at "core_dispatch.h:356".